pallets/loans/src/interest.rs

```diff
@@ -0,0 +1,119 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use primitives::{Balance, CurrencyId};
+use sp_runtime::{traits::Zero, DispatchResult};
+
+use crate::*;
+
+impl<T: Config> Pallet<T> {
+    /// Accrue interest per block and update corresponding storage
+    pub(crate) fn accrue_interest() -> DispatchResult {
+        for (currency_id, market) in Self::active_markets() {
+            let total_cash = Self::get_total_cash(currency_id);
+            let total_borrows = Self::total_borrows(currency_id);
+            let total_reserves = Self::total_reserves(currency_id);
+            let util = Self::calc_utilization_ratio(total_cash, total_borrows, total_reserves)?;
+
+            let borrow_rate = market
+                .rate_model
+                .get_borrow_rate(util)
+                .ok_or(ArithmeticError::Overflow)?;
+            let supply_rate =
+                InterestRateModel::get_supply_rate(borrow_rate, util, market.reserve_factor);
+
+            UtilizationRatio::<T>::insert(currency_id, util);
+            BorrowRate::<T>::insert(currency_id, &borrow_rate);
+            SupplyRate::<T>::insert(currency_id, supply_rate);
+
+            Self::update_borrow_index(borrow_rate, currency_id, &market)?;
+            Self::update_exchange_rate(currency_id)?;
+        }
+
+        Ok(())
+    }
+
+    /// Update the exchange rate according to the totalCash, totalBorrows and totalSupply.
+    ///
+    /// exchangeRate = (totalCash + totalBorrows - totalReserves) / totalSupply
+    pub(crate) fn update_exchange_rate(currency_id: CurrencyId) -> DispatchResult {
```

Comment:
maybe we can add https://substrate.dev/rustdocs/latest/frame_support/attr.require_transactional.html